src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java

```diff
@@ -399,6 +405,20 @@ static PrivateKey privateKey(Session session, long keyID, String algorithm,
         boolean keySensitive = (attrs[0].getBoolean() ||
                 attrs[1].getBoolean() || !attrs[2].getBoolean());
 
+        if (keySensitive && (SunPKCS11.mysunpkcs11 != null) && "RSA".equals(algorithm)) {
```

Comment:
This is a port from JDK17 PR https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/118 into JDKnext.

In JDK 17, before the PR https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/118, the codes as below. 

        if (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) {
             return new P11PrivateKey
                 (session, keyID, algorithm, keyLength, attributes);
         }

When (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) is true, then “return new P11PrivateKey (session, keyID, algorithm, keyLength, attributes);”.

After PR https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/118, it added the FIPS key wrap code between the above “if” and “return new P11PrivateKey”, so the codes as below:

        if (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) {
             if ((SunPKCS11.mysunpkcs11 != null) && "RSA".equals(algorithm)) {
                 try {
                     byte[] key = SunPKCS11.mysunpkcs11.exportKey(session.id(), attributes, keyID);
                     RSAPrivateKey rsaPrivKey = RSAPrivateCrtKeyImpl.newKey(KeyType.RSA, "PKCS#8", key);
                     if (rsaPrivKey instanceof RSAPrivateCrtKeyImpl privImpl) {
                         return new P11RSAPrivateKeyFIPS(session, keyID, algorithm, keyLength, attributes, privImpl);
                     } else {
                         return new P11RSAPrivateNonCRTKeyFIPS(session, keyID, algorithm, keyLength, attributes, rsaPrivKey);
                     }
                 } catch (PKCS11Exception | InvalidKeyException e) {
                     // Attempt failed, create a P11PrivateKey object.
                 }
             }
             return new P11PrivateKey
                 (session, keyID, algorithm, keyLength, attributes);
         }

If (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) is true, then go into the FIPS key wrap code, but if get PKCS11Exception, then “return new P11PrivateKey (session, keyID, algorithm, keyLength, attributes);” 

And these are the codes we want to port from JDK17 into JDKnext.

In JDKnext codes, the structure of this part of code is changed. It add a variable “boolean keySensitive = (attrs[0].getBoolean() || attrs[1].getBoolean() || !attrs[2].getBoolean());”. And in the “static class P11RSAPrivateKeyInternal extends P11PrivateKey”, If the “keySensitive” is true, then it will return the "P11PrivateKey". Codes as below:

            if (keySensitive) {
                return new P11RSAPrivateKeyInternal(session, keyID, algorithm,
                        keyLength, attrs);
            }

        private P11RSAPrivateKeyInternal(Session session, long keyID,
                String algorithm, int keyLength, CK_ATTRIBUTE[] attrs) {
            super(session, keyID, algorithm, keyLength, attrs);
        }

And since this PR is a port of PR https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/118 into JDKnext, so I added the FIPS key wrap code here, if (keySensitive && (SunPKCS11.mysunpkcs11 != null) && "RSA".equals(algorithm)) is true, then go into the FIPS key wrap code. But if get PKCS11Exception, it will continue go into the switch case, and because the algorithm is RSA, it will go into the the “static class P11RSAPrivateKeyInternal extends P11PrivateKey”, and because the “keySensitive” is true, then it will “return new P11PrivateKey (session, keyID, algorithm, keyLength, attrs);”.

This will keep the same logic of JDK17 and also keep the same logic of JDKnext. That is, if the “keySensitive” is true, then it will “return new P11PrivateKey (session, keyID, algorithm, keyLength, attrs);”.

Please help to review.

Reply:
The essence of my question is why jdk17 does **_not_** consider `attributes[0]` (`CKA_TOKEN`), while this version _**does**_ (in computing `keySensitive`). You haven't answered that question that I can see.

Reply:
@keithc-ca @alon-sh 

For why jdk17 does not consider attributes[0] (CKA_TOKEN), while this version does (in computing keySensitive).

From the P11Key.java changes history, I can see this “keySensitive” is added by the following commit on Dec 6, 2021, started from JDK18.

https://github.com/ibmruntimes/openj9-openjdk-jdk18/commit/df1a94ee4b26a3680a974ca8a3bdb2b71f629137

And this commit is for fixing the bug 8271566, bug details https://bugs.openjdk.org/browse/JDK-8271566

Reply:
Those links are informative, but don't answer the question of whether the change represents an issue for the related change in jdk17. Should the condition in https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java#L388 be updated to  test `attributes[0].getBoolean()` as well.

Reply:
This new check was added to reflect that in NSS PKCS11 impl. token objects which are keys are also sensitive. see https://github.com/openjdk/jdk/pull/4961#issue-958445710

This is an update that is not relevant to our purposes of import/export crypto. keys in FIPS mode as we are concerned with keys who are session objects, not token objects. (i.e. attributes[0].getBoolean() is always false for our use case). Our use case is: Liberty's LTPA code generates an RSA key and then needs to manipulate it so we need to extract that key, this key is always a session object. 

Therefore we don't need to go back and update older jdk levels to support extracting keys which are token objects.

Reply:
Just because we know one use-case for this code that doesn't care about 'token' objects, doesn't mean we should assume that all other use-cases will not care. Please open a change for jdk17 to make it consistent.

Reply:
The way I read JDK-8271566, they simply just decided not to backport to jdk17.

Reply:
@keithc-ca Then, you suggest, we just add the token object “attrs[0].getBoolean()” check into the "if" condition back to jdk17, or add all the codes in JDK-8271566 back to jdk17?

@alon-sh any advice?

Reply:
Yes, I think adding `attrs[0].getBoolean()` to the test in jdk17 is sufficient so the same conditions apply for special handling of importing or exporting keys in both jdk17 and here. If I missed some reason that's not appropriate or advisable, please explain.

Reply:
I agree it would be good to get that change backported upstream, but I don't see how the change to jdk17 creates problems: we would essentially be applying the same change there as proposed here.

Reply:
@taoliult https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/606 addresses my concern for jdk8; can you create pull requests for jdk11 and jdk17 to include `CKA_TOKEN` in the condition for special treatment in FIPS mode? Then, if this is merge, all versions will be consistent. Thanks.

Reply:
@keithc-ca Let me check with Alon to see if there is any concern.
@alon-sh For include the CKA_TOKEN in the condition for jdk11 and jdk17, any concern?

Reply:
@keithc-ca 

According to the PR ibmruntimes/openj9-openjdk-jdk8#606, create the PRs for JDK11 and JDK17, to include CKA_TOKEN in the condition for export/import plain keys in FIPS mode.

JDK11 PR: https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/606
JDK17 PR: https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/170